src/context/blocked.rs

```diff
@@ -1,18 +1,181 @@
+    if module_private_data.is_null() {
+        panic!("[auth_reply_wrapper] Module private data is null; this should not happen!");
+    }
+
+    let user_private_data = &*(module_private_data as *const BlockedClientPrivateData<T>);
+
+    let cb = match user_private_data.reply_callback.as_ref() {
+        Some(ReplyCallback::Auth(cb)) => cb,
+        None => panic!("[auth_reply_wrapper] Reply callback is null; this should not happen!"),
+    };
+
+    let data_ref = user_private_data.data.as_deref();
+
+    match cb(&context, username, password, data_ref) {
+        Ok(result) => result,
+        Err(error) => {
+            let error_msg = ValkeyString::create_and_retain(&error.to_string());
+            *err = error_msg.inner;
+            AUTH_HANDLED
+        }
+    }
+}
+
+#[allow(dead_code)]
+unsafe extern "C" fn free_callback_wrapper<T: 'static>(
+    ctx: *mut raw::RedisModuleCtx,
+    module_private_data: *mut c_void,
+) {
+    let context = Context::new(ctx);
+
+    if module_private_data.is_null() {
+        panic!("[free_callback_wrapper] Module private data is null; this should not happen!");
+    }
+
+    let user_private_data = Box::from_raw(module_private_data as *mut BlockedClientPrivateData<T>);
+
+    // Execute free_callback only if both callback and data exist
+    // Note: free_callback can exist without data - this is a valid state
+    if let Some(free_cb) = user_private_data.free_callback {
+        if let Some(data) = user_private_data.data {
+            free_cb(&context, *data);
+        }
+    }
 }
 
 // We need to be able to send the inner pointer to another thread
-unsafe impl Send for BlockedClient {}
+unsafe impl<T> Send for BlockedClient<T> {}
 
-impl Drop for BlockedClient {
+impl<T> BlockedClient<T> {
+    pub(crate) fn new(inner: *mut raw::RedisModuleBlockedClient) -> Self {
+        Self {
+            inner,
+            reply_callback: None,
+            free_callback: None,
+            data: None,
+        }
+    }
+
+    #[allow(dead_code)]
+    pub(crate) fn with_auth_callback(
+        inner: *mut raw::RedisModuleBlockedClient,
+        auth_reply_callback: fn(
+            &Context,
+            ValkeyString,
+            ValkeyString,
+            Option<&T>,
+        ) -> Result<c_int, ValkeyError>,
+        free_callback: Option<FreePrivateDataCallback<T>>,
+    ) -> Self
+    where
+        T: 'static,
+    {
+        Self {
+            inner,
+            reply_callback: Some(ReplyCallback::Auth(auth_reply_callback)),
+            free_callback,
+            data: None,
+        }
+    }
+
+    /// Sets private data for the blocked client.
+    ///
+    /// # Panics
+    /// This method will panic if called without first setting a free callback.
+    /// The free callback is required to properly clean up any resources associated
+    /// with the private data.
+    ///
+    /// # Arguments
+    /// * `data` - The private data to store
+    pub fn set_blocked_private_data(&mut self, data: T) {
```

Comment:
`
    pub fn set_blocked_private_data(&mut self, data: T) -> Result<> {
    if self.free_callback.is_none() {
       return Err("Cannot set private data without a free callback - this would leak memory")
   }
}

`